### ISSUE TITLE
How to use it in encoder-decoder model

### ISSUE THREAD

**Issue by luoda888** [gingasan/CreAT#1]
It's my code,
      logits = modeling_outputs.logits
      input_ids = self.batch["input_ids"]
      attention_mask = self.batch["attention_mask"]
      decoder_attention_mask = self.batch["decoder_attention_mask"]
      inputs_embeds = self.encoder.embed_tokens(input_ids)
      extended_input_mask = decoder_attention_mask.view(-1, decoder_attention_mask.size(-1)).unsqueeze(-1)
      ctxr = modeling_outputs.decoder_hidden_states[-1] * extended_input_mask
      delta = torch.randn_like(inputs_embeds, requires_grad=True) * self.adv_init_var

      for adv_step in range(self.adv_step):
          inputs_embeds = inputs_embeds + delta
          with autocast(enabled=self.config.fp16):
              batch = {
                  "inputs_embeds": inputs_embeds,
                  "attention_mask": attention_mask,
                  "input_ids": None,
                  "labels": self.batch["labels"],
                  "output_hidden_states": True,
              }

              adv_modeling_outputs = self.model(**batch)
              loss_ptb = adv_modeling_outputs.loss
              logits_ptb = adv_modeling_outputs.logits
              ctxr_ptb = adv_modeling_outputs.decoder_hidden_states[-1] * extended_input_mask

          if adv_step == self.adv_step - 1:
              break

          loss_ptb = loss_ptb - cos_loss(ctxr_ptb, ctxr.detach()) * self.adv_temp
          delta = self._inner_update(delta, loss_ptb)
          delta = delta.requires_grad_()

**Comment by gingasan** [comment]
Yes. Our method is agnostic if the model has a decoder (e.g. T5, GPT). The only thing is to modify the input part of the code.
Your code is right. The only problem is here: inputs_embeds = inputs_embeds + delta. This line should be removed. The 'delta' should be added in below: "inputs_embeds": inputs_embeds + delta. If you need more adv_steps (>1), the perturbations should not be accumulated.

You can refer to R3F paper and their code, where encoder-decoder models are experimented.
BETTER FINE-TUNING BY REDUCING REPRESENTATIONAL COLLAPSE

----- 原始邮件 -----
发件人: "Robin Lee" ***@***.***>
收件人: "gingasan/CreAT" ***@***.***>
抄送: "Subscribed" ***@***.***>
发送时间: 星期二, 2023年 5 月 16日 上午 10:21:00
主题: [gingasan/CreAT] How to use it in encoder-decoder model (Issue #1)

It's my code,
      logits = modeling_outputs.logits
      input_ids = self.batch["input_ids"]
      attention_mask = self.batch["attention_mask"]
      decoder_attention_mask = self.batch["decoder_attention_mask"]
      inputs_embeds = self.encoder.embed_tokens(input_ids)
      extended_input_mask = decoder_attention_mask.view(-1, decoder_attention_mask.size(-1)).unsqueeze(-1)
      ctxr = modeling_outputs.decoder_hidden_states[-1] * extended_input_mask
      delta = torch.randn_like(inputs_embeds, requires_grad=True) * self.adv_init_var

      for adv_step in range(self.adv_step):
          inputs_embeds = inputs_embeds + delta
          with autocast(enabled=self.config.fp16):
              batch = {
                  "inputs_embeds": inputs_embeds,
                  "attention_mask": attention_mask,
                  "input_ids": None,
                  "labels": self.batch["labels"],
                  "output_hidden_states": True,
              }

              adv_modeling_outputs = self.model(**batch)
              loss_ptb = adv_modeling_outputs.loss
              logits_ptb = adv_modeling_outputs.logits
              ctxr_ptb = adv_modeling_outputs.decoder_hidden_states[-1] * extended_input_mask

          if adv_step == self.adv_step - 1:
              break

          loss_ptb = loss_ptb - cos_loss(ctxr_ptb, ctxr.detach()) * self.adv_temp
          delta = self._inner_update(delta, loss_ptb)
          delta = delta.requires_grad_()

-- 
Reply to this email directly or view it on GitHub:
https://github.com/gingasan/CreAT/issues/1
